packages/x-data-grid/src/utils/createSelector.ts

```diff
@@ -32,10 +42,31 @@ type SelectorArgs<Selectors extends ReadonlyArray<Selector<any>>, Result> =
   // Input selectors as separate inline arguments
   | [...Selectors, (...args: SelectorResultArray<Selectors>) => Result];
 
+type SelectorResultArrayWithArgs<Selectors extends ReadonlyArray<Selector<any>>, Args> = [
+  ...SelectorResultArray<Selectors>,
+  Args,
+];
+
+type ArgumentsSelectorArgs<Selectors extends ReadonlyArray<Selector<any>>, Args, Result> =
+  // Input selectors as a separate array
+  | [
+      selectors: [...Selectors],
+      combiner: (...args: SelectorResultArrayWithArgs<Selectors, Args>) => Result,
+    ]
+  // Input selectors as separate inline arguments
+  | [...Selectors, (...args: SelectorResultArrayWithArgs<Selectors, Args>) => Result];
+
 type CreateSelectorFunction = <Selectors extends ReadonlyArray<Selector<any>>, Result>(
   ...items: SelectorArgs<Selectors, Result>
 ) => OutputSelector<StateFromSelectorList<Selectors>, Result>;
 
+type CreateArgumentsSelectorFunction = <Args = void>() => <
```

Comment:
Two-step generics are used to provide a better type coverage for the provided arguments. We first define `Args` separately with an outer function, allowing the user to specify `Args` as explicit arguments. The inner function has the inferred values (Selectors and Result) since I couldn't find a way in typescript to combine explicit and inferred function parameters, something like:
```tsx
  type CreateArgumentsSelectorFunction = <Args, Selectors extends ReadonlyArray<Selector<any>>, Result>(
    ...items: ArgumentsSelectorArgs<Selectors, Args, Result>
  ) => OutputArgumentsSelector<StateFromSelectorList<Selectors>, Args, Result>;
```

So that the user could use it as follows without messing up the type inference for argument selectors:
```tsx
  const someSelector = createSelector<Args>(selector1, selector2, (output1, output2, args /* of type Args */) => result)
```

Now the user has to use an additional parenthesis for selectors with arguments. Since the selectors are static, this better types coverage/dx seems a fair trade-off.

```tsx
const someSelector = createSelector<Args>()(selector1, selector2, (output1, output2, args /* of type Args */) => result)
```

Thoughts?

Reply:
should `createArgumentsSelector` also be `createSelectorV8` that will replace current `createSelector` in v8
arguments can be optional to support the current creators.

Reply:
personally, I would rather do some additional updates on TS side (even use `any` for now) than have an additional `()` for each selector creation

Reply:
> should createArgumentsSelector also be createSelectorV8 that will replace current createSelector in v8
arguments can be optional to support the current creators.

I initially [aimed](https://github.com/mui/mui-x/pull/13757/files#r1700571749) this. But using this approach, it's not possible to type the `Args` properly which might yield a bad DX.

If we aim it with the current approach, we'll have to add an extra parenthesis next to each selector (including the existing ones) in v8 which is definitely out of the question.

Here's a short comparison of the approaches we could follow:

## Options that we have:

### Have separate selector creators for selectors with arguments

This is the current approach being followed in this PR.
Following are the arguments that made me think it's a better approach.
1. Better DX with the extra parenthesis trade off. (Which IMO is not a big problem because it is only needed at the time of selector creation, not the use)
2. Only a small subset of (mostly new) selectors that require some runtime arguments will be affected , the others will be using the same `createSelector` and `createSelectorMemoized`.
3. There are no performance implications of the extra parenthesis since it's run statically.
4. The extra learning curve introduced with the extra parenthesis could be catered to by adding a nice documentation section. (Though this shouldn't be a big problem as most of the times, the selectors are created internally)

### Same selector creators while trading it off with bad DX

Introduce the arguments functionality in the existing `createSelector` and `createSelectorMemoized`, and create new variants that will replace them in v8.
1. No typescript support. `Args` will be `any` and the users will always have to either look over the JSDoc or the selector code to discover what arguments it expects (if at all).
2. This type of manual discoverability adds a mental burden to the user's mind and increases the chances of errors.
3. The positive point is that the API will be the same, so no learning curve, and of course no extra parenthesis.

I am open to disagreements, but I personally feel that the DX improvement we achieve using the first approach weighs out the extra needed parenthesis. 🤷

Reply:
I don't mind the `()` if it gets us nice typescript. This is mostly for internal use or for very advanced use-cases/users, we don't need to baby-proof this API.

Reply:
What do you think of this approach?

If `Args` are added to the inferred arguments you can still add the proper args types by changing this
```
export const gridDataSourceLoadingIdSelector = createArgumentsSelector<GridRowId>()(
  gridDataSourceStateSelector,
  (dataSource, id) => dataSource.loading[id] ?? false,
);
```
to this
```
export const gridDataSourceLoadingIdSelector = createArgumentsSelector(
  gridDataSourceStateSelector,
  (dataSource, id: GridRowId) => dataSource.loading[id] ?? false,
);
```

Reply:
How could I miss this. 🤦
Updated: https://github.com/mui/mui-x/pull/14236/commits/99753bc1ff4fd2446ae02361802d794cf1c5a8b6
It's simpler and consistent with the existing API now.
Thank you